File: EclipsisCommon/src/main/java/noppes/npcs/objects/items/ItemScripted.java
package noppes.npcs.objects.items;

import net.minecraft.entity.EntityLivingBase;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.math.MathHelper;
import noppes.npcs.api.NpcAPI;
import noppes.npcs.api.item.IItemStack;
import noppes.npcs.api.wrapper.ItemScriptedWrapper;
import noppes.npcs.constants.EnumPacketServer;
import noppes.npcs.objects.CreativeTabNpcs;
import noppes.npcs.util.IPermission;

import java.util.HashMap;
import java.util.Map;

public class ItemScripted extends Item implements IPermission {
    public static Map<Integer, String> Resources = new HashMap<>();

    public ItemScripted() {
        maxStackSize = 1;
        setCreativeTab(CreativeTabNpcs.INSTANCE);
        setHasSubtypes(true);
    }

    @Override
    public Item setTranslationKey(String name) {
        super.setTranslationKey(name);
        setRegistryName(new ResourceLocation("customnpcs", name));
        return this;
    }

    @Override
    public boolean isAllowed(EnumPacketServer e) {
        return e == EnumPacketServer.ScriptItemDataGet || e == EnumPacketServer.ScriptItemDataSave;
    }

    public static ItemScriptedWrapper GetWrapper(ItemStack stack) {
        return (ItemScriptedWrapper) NpcAPI.Instance().getIItemStack(stack);
    }

    @Override
    public boolean showDurabilityBar(ItemStack stack) {
        IItemStack istack = NpcAPI.Instance().getIItemStack(stack);
        if (istack instanceof ItemScriptedWrapper)
            return ((ItemScriptedWrapper) istack).durabilityShow;
        return super.showDurabilityBar(stack);
    }

    @Override
    public double getDurabilityForDisplay(ItemStack stack) {
        IItemStack istack = NpcAPI.Instance().getIItemStack(stack);
        if (istack instanceof ItemScriptedWrapper)
            return 1 - ((ItemScriptedWrapper) istack).durabilityValue;
        return super.getDurabilityForDisplay(stack);
    }

    @Override
    public int getRGBDurabilityForDisplay(ItemStack stack) {
        IItemStack istack = NpcAPI.Instance().getIItemStack(stack);
        if (!(istack instanceof ItemScriptedWrapper))
            return super.getRGBDurabilityForDisplay(stack);
        int color = ((ItemScriptedWrapper) istack).durabilityColor;
        if (color >= 0)
            return color;
        return MathHelper.hsvToRGB(Math.max(0.0F, (float) (1.0F - getDurabilityForDisplay(stack))) / 3.0F, 1.0F, 1.0F);
    }

    @Override
    public int getItemStackLimit(ItemStack stack) {
        IItemStack istack = NpcAPI.Instance().getIItemStack(stack);
        if (istack instanceof ItemScriptedWrapper)
            return istack.getMaxStackSize();
        return super.getItemStackLimit(stack);
    }

    @Override
    public boolean hitEntity(ItemStack stack, EntityLivingBase target, EntityLivingBase attacker) {
        return true;
    }
}

File: EclipsisCommon/src/main/java/noppes/npcs/common/objects/items/ItemScripted.kt
package noppes.npcs.common.objects.items

import com.teamwizardry.librarianlib.features.base.item.ItemMod
import net.minecraft.entity.EntityLivingBase
import net.minecraft.item.ItemStack
import net.minecraft.util.math.MathHelper
import noppes.npcs.api.NpcAPI
import noppes.npcs.api.wrapper.ItemScriptedWrapper
import noppes.npcs.common.objects.CreativeTabNpcs
import noppes.npcs.constants.EnumPacketServer
import noppes.npcs.util.IPermission
import ru.glassspirit.eclipsis.kotlin.setModCreativeTab
import java.util.*

class ItemScripted : ItemMod("scripted_item"), IPermission {
    init {
        maxStackSize = 1
        setHasSubtypes(true)
        setFull3D()
        this.setModCreativeTab(CreativeTabNpcs)
    }

    override fun isAllowed(e: EnumPacketServer): Boolean {
        return e == EnumPacketServer.ScriptItemDataGet || e == EnumPacketServer.ScriptItemDataSave
    }

    override fun showDurabilityBar(stack: ItemStack): Boolean {
        val istack = NpcAPI.instance()!!.getIItemStack(stack)
        return (istack as? ItemScriptedWrapper)?.durabilityShow ?: super.showDurabilityBar(stack)
    }

    override fun getDurabilityForDisplay(stack: ItemStack): Double {
        val istack = NpcAPI.instance()!!.getIItemStack(stack)
        return if (istack is ItemScriptedWrapper) 1 - istack.durabilityValue else super.getDurabilityForDisplay(stack)
    }

    override fun getRGBDurabilityForDisplay(stack: ItemStack): Int {
        val istack = NpcAPI.instance()!!.getIItemStack(stack) as? ItemScriptedWrapper ?: return super.getRGBDurabilityForDisplay(stack)
        val color = istack.durabilityColor
        return if (color >= 0) color else MathHelper.hsvToRGB(Math.max(0.0f, (1.0f - getDurabilityForDisplay(stack)).toFloat()) / 3.0f, 1.0f, 1.0f)
    }

    override fun getItemStackLimit(stack: ItemStack): Int {
        val istack = NpcAPI.instance()!!.getIItemStack(stack)
        return (istack as? ItemScriptedWrapper)?.maxStackSize ?: super.getItemStackLimit(stack)
    }

    override fun hitEntity(stack: ItemStack, target: EntityLivingBase, attacker: EntityLivingBase): Boolean {
        return true
    }

    companion object {
        var Resources: Map<Int, String> = HashMap()

        fun getWrapper(stack: ItemStack): ItemScriptedWrapper {
            return NpcAPI.instance()!!.getIItemStack(stack) as ItemScriptedWrapper
        }
    }
}
